icu4c/source/i18n/smpdtfmt.cpp

```diff
@@ -594,11 +595,20 @@ SimpleDateFormat& SimpleDateFormat::operator=(const SimpleDateFormat& other)
     fLocale = other.fLocale;
 
     // TimeZoneFormat can now be set independently via setter.
-    // If it is NULL, it will be lazily initialized from locale
+    // If it is NULL, it will be lazily initialized from locale.
+    //
+    // This lazy initialization may happen from const functions,
+    // meaning it could happen concurrently on other.fTimeZoneFormat
+    // in the following code. Thus requiring synchronization.
     delete fTimeZoneFormat;
-    fTimeZoneFormat = NULL;
-    if (other.fTimeZoneFormat) {
-        fTimeZoneFormat = new TimeZoneFormat(*other.fTimeZoneFormat);
+    fTimeZoneFormat = nullptr;
+    TimeZoneFormat *otherTZFormat;
+    {
+        Mutex m(&LOCK);
+        otherTZFormat = other.fTimeZoneFormat;
+    }
+    if (otherTZFormat) {
+        fTimeZoneFormat = new TimeZoneFormat(*otherTZFormat);
```

Comment:
I had to think through your change and comment here...

I think the main point is that this pointer is lazy-initialized, so in normal, valid use there could be a race from nullptr to pointer to object, but not a race in transition *to* nullptr. That is, if you grab the pointer and it's not nullptr you need not remain inside the Mutex because the object is fully built and won't be deleted. Right?

Please beef up the comment...